src/ucp/tag/offload_proto.c

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucp/proto/proto_single.inl>
+
+static ucs_status_t
+ucp_proto_eager_tag_offload_short_progress(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_t *ep       = req->send.ep;
+    const ucp_proto_single_priv_t *spriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    status = uct_ep_tag_eager_short(ep->uct_eps[spriv->super.lane],
+                                    req->send.msg_proto.tag.tag,
+                                    req->send.state.dt_iter.type.contig.buffer,
+                                    req->send.state.dt_iter.length);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        req->send.lane = spriv->super.lane; /* for pending add */
+        return status;
+    }
+
+    ucp_datatype_iter_cleanup(&req->send.state.dt_iter,
+                              UCS_BIT(UCP_DATATYPE_CONTIG));
+
+    ucs_assert(status != UCS_INPROGRESS);
+    ucp_request_complete_send(req, status);
+
+    return UCS_OK;
+}
+
+static ucs_status_t ucp_proto_eager_tag_offload_short_init(
+        const ucp_proto_init_params_t *init_params)
+{
+    const ucp_proto_select_param_t *select_param = init_params->select_param;
+    ucp_proto_single_init_params_t params = {
+        .super.super         = *init_params,
+        .super.latency       = -150e-9, /* no extra memory access to fetch data */
+        .super.overhead      = 0,
+        .super.cfg_thresh    = UCS_MEMUNITS_AUTO,
+        .super.cfg_priority  = 0,
+        .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.tag.eager.max_short),
+        .super.hdr_size      = sizeof(ucp_tag_t),
+
+        /* Do not set UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY, because by default
+         * small messages are not posted to the HW on the receiver side and arrive
+         * unexpectedly. Thus, they are handled by the SW (UCP), which does memory
+         * copy.
+         */
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_MAX_FRAG,
+        .lane_type           = UCP_LANE_TYPE_TAG,
+        .tl_cap_flags        = UCT_IFACE_FLAG_TAG_EAGER_SHORT
+    };
+
+    if ((select_param->op_id != UCP_OP_ID_TAG_SEND) ||
+        /* offload proto can not be used if no tag offload lane configured */
+        !ucp_ep_config_key_has_tag_lane(init_params->ep_config_key) ||
+        /* short protocol requires contig/host */
+        (select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        !UCP_MEM_IS_HOST(select_param->mem_type)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
+}
+
+static ucp_proto_t ucp_eager_tag_offload_short_proto = {
+    .name       = "egr/single/offload/short",
```

Comment:
maybe just "egr/offload/short"
since offload does not have a "multi" variant right?

Reply:
fixed